### PR TITLE
hotplug_mem_migration: Add memory and migration cross-test

### PR DESCRIPTION
**Pull request by zhenyzha**
Add memory and migration cross-test scenarios,
This scenario simulates the migration of hot-inserted memory after use, and then unplugs it after the migration is complete.

ID: 2152804
Signed-off-by: zhenyzha <zhenyzha@redhat.com>

### CODE REVIEW

**Comment by zhenyzha** [comment]
```
 (1/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.one.arm64-pci: STARTED
 (1/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.one.arm64-pci: PASS (241.50 s)
 (2/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.two.arm64-pci: STARTED
 (2/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.two.arm64-pci: PASS (247.76 s)
 (3/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.one.arm64-pci: STARTED
 (3/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.one.arm64-pci: PASS (242.79 s)
 (4/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.two.arm64-pci: STARTED
 (4/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.two.arm64-pci: PASS (249.81 s)
 (5/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.two.arm64-pci: STARTED
 (5/5) Host_RHEL.m9.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.two.arm64-pci: PASS (248.61 s)
```

@YongxueHong @mcasquer Could you pls help review when you are free. thanks.

File: qemu/tests/cfg/hotplug_mem_migration.cfg [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    cmd_new_folder = ' rm -rf /tmp/numa_test/ && mkdir /tmp/numa_test'
+    numa_test = 'numactl -m %s dd if=/dev/urandom of=/tmp/numa_test/test '
+    numa_test += 'bs=1k count=2621440 && rm -rf /tmp/numa_test/'
+    stress_args = '--cpu 4 --io 4 --vm 2 --vm-bytes 2048M'
+    variants numa_nodes:
+        - one:
+            guest_numa_nodes = "node0"
+            mem_devs = "memN0"
+            use_mem_memN0 = "no"
+            size_mem_memN0 = 4096M
```
Comment (mcasquer) [inline qemu/tests/cfg/hotplug_mem_migration.cfg:41]:
```
x86_64:
    size_mem_memN0 = 8192M
```
You can include it for x86_64

File: qemu/tests/cfg/hotplug_mem_migration.cfg [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            del numa_mem
+            del numa_cpus
+            numa_nodeid = 0
+            mem_devs = "memN0 memN1"
+            numa_memdev_node0 = mem-memN0
+            numa_memdev_node1 = mem-memN1
+            use_mem_memN0 = "no"
+            use_mem_memN1 = "no"
+            size_mem_memN0 = 2048M
+            size_mem_memN1 = 2048M
```
Comment (mcasquer) [inline qemu/tests/cfg/hotplug_mem_migration.cfg:58]:
```
x86_64:
    size_mem_memN0 = 4096M
    size_mem_memN1 = 4096M
```
You can include it for x86_64

**Comment by mcasquer** [comment]
@zhenyzha after the suggested changes all the cases pass successfully:

```bash
 (1/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.one.q35: STARTED
 (1/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.one.q35: PASS (297.60 s)
 (2/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.two.q35: STARTED
 (2/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.two.q35: PASS (317.70 s)
 (3/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.one.q35: STARTED
 (3/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.one.q35: PASS (299.83 s)
 (4/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.two.q35: STARTED
 (4/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.two.q35: PASS (329.78 s)
 (5/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.two.q35: STARTED
 (5/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.two.q35: PASS (323.74 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

File: qemu/tests/cfg/hotplug_mem_migration.cfg (hunk 2 above)
Comment (zhenyzha) [reply qemu/tests/cfg/hotplug_mem_migration.cfg]: Thank a lot for your review.
Considering that the kernel will take up more and more memory in the future, 
I increased the mem size of all platforms and simultaneously increased the pressure value.

File: qemu/tests/cfg/hotplug_mem_migration.cfg [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    no Host_RHEL.6 Host_RHEL.7 Host_RHEL.8
+    no RHEL.6 RHEL.7 RHEL.8
+    no Windows
+    threshold = 0.12
+    # Notes:
+    #    The threshold on arm and ppc was confirmed with developer
+    #    since crashkernel size is much bigger than x86
+    aarch64,ppc64,ppc64le:
+        threshold = 0.15
+    sub_type = migration
```
Comment (YongxueHong) [inline qemu/tests/cfg/hotplug_mem_migration.cfg:19]:
```suggestion
```
Looks like this parameter is not used

File: qemu/tests/hotplug_mem_migration.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        LOG_JOB.info("Use the hotplug memory by numa %s.", node)
+        session.cmd(cmd_new_folder)
+        session.cmd(numa_test % re.search(r'\d+', node).group())
+    try:
+        stress_args = params.get("stress_args")
+        stress_test = utils_test.VMStress(vm, "stress", params,
+                                          stress_args=stress_args)
+        stress_test.load_stress_tool()
+    except utils_test.StressError as info:
+        raise exceptions.TestError(info)
```
Comment (YongxueHong) [inline qemu/tests/hotplug_mem_migration.py:73]:
```suggestion
        test.error(info)
```

File: qemu/tests/hotplug_mem_migration.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    time.sleep(60)
+    stress_test.unload_stress()
+    stress_test.clean()
+    # do migration
+    mig_timeout = float(params.get("mig_timeout", "1200"))
```
Comment (YongxueHong) [inline qemu/tests/hotplug_mem_migration.py:78]:
```suggestion
    mig_timeout = params.get_numeric("mig_timeout", 1200, float))
```

File: qemu/tests/hotplug_mem_migration.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    mig_protocol = params.get("migration_protocol", "tcp")
+    vm.migrate(mig_timeout, mig_protocol, env=env)
+    session = vm.wait_for_login()
```
Comment (YongxueHong) [inline qemu/tests/hotplug_mem_migration.py:81]:
```suggestion
```
Actually, the session will not disconnect after migrating, the session object will be dead if the VM reboot.

File: qemu/tests/hotplug_mem_migration.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    4) Send a migration command to the source VM and wait until it's finished
+    5) Check if memory size is correct inside guest in destination
+    6）Unplug memory device and check it
+    7) Reboot guest
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def check_online_mem(session):
```
Comment (YongxueHong) [inline qemu/tests/hotplug_mem_migration.py:30]:
```suggestion
    def _check_online_mem(session):
```

**Comment by mcasquer** [comment]
@zhenyzha apart from @YongxueHong comments, the rest looks good to me.

```bash
 (1/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.one.q35: STARTED
 (1/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.one.q35: PASS (341.46 s)
 (2/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.two.q35: STARTED
 (2/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.two.q35: PASS (372.07 s)
 (3/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.one.q35: STARTED
 (3/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.one.q35: PASS (338.69 s)
 (4/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.two.q35: STARTED
 (4/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.two.q35: PASS (372.50 s)
 (5/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.two.q35: STARTED
 (5/5) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.two.q35: PASS (375.81 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
```

**Review by mcasquer — DISMISSED** [review]
ACK

File: qemu/tests/cfg/hotplug_mem_migration.cfg [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    ping_pong = 1
+    migration_protocol = "tcp"
+    target_mems = "plug1 plug2"
+    cmd_check_online_mem = lsmem
+    cmd_new_folder = ' rm -rf /tmp/numa_test/ && mkdir /tmp/numa_test'
+    numa_test = 'numactl -m %s dd if=/dev/urandom of=/tmp/numa_test/test '
+    numa_test += 'bs=1k count=5242880 && rm -rf /tmp/numa_test/'
+    stress_args = '--cpu 4 --io 4 --vm 2 --vm-bytes 4096M'
+    variants numa_nodes:
+        - one:
```
[the remark this answers is not in the document]
Comment (zhenyzha) [reply qemu/tests/cfg/hotplug_mem_migration.cfg]: Thank a lot for your review :)
has been corrected .

File: qemu/tests/cfg/hotplug_mem_migration.cfg [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            guest_numa_nodes = "node0"
+            mem_devs = "memN0"
+            use_mem_memN0 = "no"
+            size_mem_memN0 = 8192M
+            backend_mem_memN0 = memory-backend-ram
+            numa_memdev_node0 = mem-memN0
+            del numa_mem
+            del numa_cpus
+            del numa_nodeid
+        - two:
```
[the remark this answers is not in the document]
Comment (zhenyzha) [reply qemu/tests/cfg/hotplug_mem_migration.cfg]: Corrected

File: qemu/tests/cfg/hotplug_mem_migration.cfg (hunk 3 above)
Comment (zhenyzha) [reply qemu/tests/cfg/hotplug_mem_migration.cfg]: Corrected

File: qemu/tests/hotplug_mem_migration.py (hunk 4 above)
Comment (zhenyzha) [reply qemu/tests/hotplug_mem_migration.py]: Corrected

File: qemu/tests/hotplug_mem_migration.py (hunk 5 above)
Comment (zhenyzha) [reply qemu/tests/hotplug_mem_migration.py]: Corrected

File: qemu/tests/hotplug_mem_migration.py (hunk 6 above)
Comment (zhenyzha) [reply qemu/tests/hotplug_mem_migration.py]: Corrected

File: qemu/tests/hotplug_mem_migration.py (hunk 7 above)
Comment (zhenyzha) [reply qemu/tests/hotplug_mem_migration.py]: Corrected

**Comment by zhenyzha** [comment]
@YongxueHong hello, the patch has been updated.
Could you pls help review again when you are free. thanks.

git am 3566.patch
Applying: hotplug_mem_migration: Add memory and migration cross-test

```
 (1/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.with_single_numa_node.arm64-pci: STARTED
 (1/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.with_single_numa_node.arm64-pci: PASS (299.27 s)
 (2/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.with_multi_numa_nodes.arm64-pci: STARTED
 (2/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_ram.with_multi_numa_nodes.arm64-pci: PASS (323.04 s)
 (3/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.with_single_numa_node.arm64-pci: STARTED
 (3/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.with_single_numa_node.arm64-pci: PASS (297.83 s)
 (4/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.with_multi_numa_nodes.arm64-pci: STARTED
 (4/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_file.with_multi_numa_nodes.arm64-pci: PASS (318.17 s)
 (5/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.with_multi_numa_nodes.arm64-pci: STARTED
 (5/5) Host_RHEL.m9.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.hotplug_mem_migration.backend_memfd.with_multi_numa_nodes.arm64-pci: PASS (324.67 s)
```

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Pull request merged**